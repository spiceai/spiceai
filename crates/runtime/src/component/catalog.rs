/*
Copyright 2024 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use globset::{Glob, GlobSet, GlobSetBuilder};
use snafu::prelude::*;
use spicepod::component::{catalog as spicepod_catalog, params::Params};
use std::collections::HashMap;

#[derive(Debug, Clone)]
pub struct Catalog {
    pub provider: String,
    pub catalog_id: Option<String>,
    pub name: String,
    orig_include: Vec<String>,
    pub include: Option<GlobSet>,
    pub params: HashMap<String, String>,
    pub dataset_params: HashMap<String, String>,
}

impl PartialEq for Catalog {
    fn eq(&self, other: &Self) -> bool {
        self.provider == other.provider
            && self.catalog_id == other.catalog_id
            && self.name == other.name
            && self.orig_include == other.orig_include
            && self.params == other.params
            && self.dataset_params == other.dataset_params
    }
}

impl TryFrom<spicepod_catalog::Catalog> for Catalog {
    type Error = crate::Error;

    fn try_from(catalog: spicepod_catalog::Catalog) -> std::result::Result<Self, Self::Error> {
        let provider = Catalog::provider(&catalog.from);
        let catalog_id = Catalog::catalog_id(&catalog.from).map(String::from);

        let mut globset_opt: Option<GlobSet> = None;
        if !catalog.include.is_empty() {
            let mut globset_builder = GlobSetBuilder::new();
            let include_iter = catalog.include.iter().map(|pattern| {
                Glob::new(pattern).context(crate::InvalidGlobPatternSnafu { pattern })
            });
            for glob in include_iter {
                globset_builder.add(glob?);
            }

            globset_opt = Some(
                globset_builder
                    .build()
                    .context(crate::ErrorConvertingGlobSetToRegexSnafu)?,
            );
        }

        Ok(Catalog {
            provider: provider.to_string(),
            catalog_id,
            name: catalog.name,
            orig_include: catalog.include.clone(),
            include: globset_opt,
            params: catalog
                .params
                .as_ref()
                .map(Params::as_string_map)
                .unwrap_or_default(),
            dataset_params: catalog
                .dataset_params
                .as_ref()
                .map(Params::as_string_map)
                .unwrap_or_default(),
        })
    }
}

impl Catalog {
    pub fn new(from: &str, name: &str) -> Self {
        Catalog {
            provider: Catalog::provider(from).to_string(),
            catalog_id: Catalog::catalog_id(from).map(String::from),
            name: name.into(),
            orig_include: Vec::default(),
            include: None,
            params: HashMap::default(),
            dataset_params: HashMap::default(),
        }
    }

    /// Returns the catalog provider - the first part of the `from` field before the first `:`.
    ///
    /// # Examples
    ///
    /// ```
    /// use runtime::component::catalog::Catalog;
    ///
    /// let catalog = Catalog::new("foo:bar", "bar");
    ///
    /// assert_eq!(catalog.provider, "foo".to_string());
    /// ```
    ///
    /// ```
    /// use runtime::component::catalog::Catalog;
    ///
    /// let catalog = Catalog::new("foo", "bar");
    ///
    /// assert_eq!(catalog.provider, "foo".to_string());
    /// ```
    #[must_use]
    fn provider(from: &str) -> &str {
        from.split(':').next().unwrap_or(from)
    }

    /// Returns the catalog id - the second part of the `from` field after the first `:`.
    /// This is optional and will return the default catalog from the provider if not set.
    ///
    /// # Examples
    ///
    /// ```
    /// use runtime::component::catalog::Catalog;
    ///
    /// let catalog = Catalog::new("foo:bar", "bar");
    ///
    /// assert_eq!(catalog.catalog_id, Some("bar".to_string()));
    /// ```
    ///
    /// ```
    /// use runtime::component::catalog::Catalog;
    ///
    /// let catalog = Catalog::new("foo", "bar");
    ///
    /// assert_eq!(catalog.catalog_id, None);
    /// ```
    #[must_use]
    fn catalog_id(from: &str) -> Option<&str> {
        match from.find(':') {
            Some(index) => Some(&from[index + 1..]),
            None => None,
        }
    }
}
