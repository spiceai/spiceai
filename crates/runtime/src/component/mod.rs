pub mod catalog;
pub mod dataset;
pub mod view;
