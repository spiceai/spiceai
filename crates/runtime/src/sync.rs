use snafu::prelude::*;

#[derive(Debug, Snafu)]
#[allow(clippy::enum_variant_names)]
pub enum Error {
    #[snafu(display("Invalid length"))]
    InvalidLength { context: &'static str },

    #[snafu(display("Invalid magic bytes"))]
    InvalidMagicBytes,

    #[snafu(display("Invalid prefix"))]
    InvalidPrefix { source: std::string::FromUtf8Error },

    #[snafu(display("Invalid LSN"))]
    InvalidLogSequenceNumber,
}

type Result<T, E = Error> = std::result::Result<T, E>;

pub mod cursor {
    use super::*;

    const MAGIC_BYTES: [u8; 4] = [83, 80, 83, 67]; // ASCII for "SPSC" (SPice Sync Cursor)

    // Cursor is used to synchronize data between the
    // Spice.ai Runtime and the Spice.ai CLI.
    pub struct Cursor {
        version: u8,
        prefix: String,
        log_sequence_number: u64,
    }

    impl Cursor {
        pub fn parse(cursor: impl AsRef<[u8]>) -> Result<Self> {
            let cursor = cursor.as_ref();
            let mut iter = cursor.iter();

            // Check for magic bytes
            for &magic_byte in &MAGIC_BYTES {
                if iter.next() != Some(&magic_byte) {
                    return InvalidMagicBytesSnafu.fail();
                }
            }

            // Parse version
            let version = *iter.next().ok_or(Error::InvalidLength {
                context: "version: *iter.next().ok_or",
            })?;

            // Parse ASCII string prefix
            let prefix_length = *iter.next().ok_or(Error::InvalidLength {
                context: "prefix_length: *iter.next().ok_or",
            })?;
            let prefix_bytes: Vec<u8> = iter.by_ref().take(prefix_length.into()).copied().collect();
            let prefix = String::from_utf8(prefix_bytes)
                .context(InvalidPrefixSnafu)?
                .trim_matches(char::is_whitespace)
                .to_string();

            // Parse log sequence number (8 bytes, big endian)
            let log_sequence_number_bytes: Vec<u8> = iter.by_ref().take(8).copied().collect();
            if log_sequence_number_bytes.len() != 8 {
                return InvalidLengthSnafu {
                    context: "log_sequence_number_bytes.len() != 8",
                }
                .fail();
            }
            let log_sequence_number_bytes: [u8; 8] = match log_sequence_number_bytes.try_into() {
                Ok(bytes) => bytes,
                Err(_) => return InvalidLogSequenceNumberSnafu.fail(),
            };
            let log_sequence_number = u64::from_be_bytes(log_sequence_number_bytes);

            Ok(Cursor {
                version,
                prefix,
                log_sequence_number,
            })
        }

        pub fn to_bytes(&self) -> Result<Vec<u8>> {
            let mut bytes = Vec::with_capacity(28);

            // Add magic bytes
            bytes.extend_from_slice(&MAGIC_BYTES);

            // Add version byte
            bytes.push(self.version);

            // Add ASCII string prefix
            let prefix_bytes = self.prefix.as_bytes().to_vec();
            let prefix_length = match u8::try_from(prefix_bytes.len()) {
                Ok(length) => length,
                Err(_err) => {
                    return Err(Error::InvalidLength {
                        context: "prefix length > u8::MAX (255)",
                    })
                }
            };
            bytes.push(prefix_length);
            bytes.extend_from_slice(&prefix_bytes);

            // Add log sequence number (8 bytes, big endian)
            let log_sequence_number_bytes = self.log_sequence_number.to_be_bytes();
            bytes.extend_from_slice(&log_sequence_number_bytes);

            Ok(bytes)
        }
    }

    #[cfg(test)]
    #[allow(clippy::expect_used)]
    mod tests {
        use super::*;

        #[test]
        fn test_parse() {
            let cursor_test_bytes: Vec<u8> = vec![
                83, 80, 83, 67, // Magic bytes (SPSC)
                1,  // Version
                12, 115, 112, 105, 99, 101, 46, 97, 105, 46, 99, 108, 105, // Prefix
                255, 255, 255, 255, 255, 255, 0, 0, // Log sequence number
            ];
            let cursor = match Cursor::parse(&cursor_test_bytes) {
                Ok(cursor) => cursor,
                Err(err) => panic!("Unable to parse cursor: {err:?}"),
            };
            assert_eq!(cursor.version, 1);
            assert_eq!(cursor.prefix, "spice.ai.cli");
            assert_eq!(cursor.log_sequence_number, 0xFFFF_FFFF_FFFF_0000_u64);

            let cursor_bytes = cursor
                .to_bytes()
                .expect("Unable to convert cursor to bytes");

            assert_eq!(cursor_bytes, cursor_test_bytes);
        }
    }
}
