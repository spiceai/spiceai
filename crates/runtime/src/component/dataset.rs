/*
Copyright 2024 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use data_components::util::column_reference;
use datafusion::sql::TableReference;
use snafu::prelude::*;
use spicepod::component::{
    dataset as spicepod_dataset, embeddings::ColumnEmbeddingConfig, params::Params,
};
use std::{collections::HashMap, time::Duration};

#[derive(Debug, Snafu)]
pub enum Error {
    #[snafu(display(
        "Column for index {index} not found in schema. Valid columns: {valid_columns}"
    ))]
    IndexColumnNotFound {
        index: String,
        valid_columns: String,
    },

    #[snafu(display("Unable to get table constraints: {source}"))]
    UnableToGetTableConstraints {
        source: datafusion::error::DataFusionError,
    },

    #[snafu(display("Unable to convert a SchemaRef to a DFSchema: {source}"))]
    UnableToConvertSchemaRefToDFSchema {
        source: datafusion::error::DataFusionError,
    },

    #[snafu(display("Only one `on_conflict` target can be specified, or all `on_conflict` targets must be specified and set to `drop`. {extra_detail}"))]
    OnConflictTargetMismatch { extra_detail: String },

    #[snafu(display("Error parsing column reference {column_ref}: {source}"))]
    UnableToParseColumnReference {
        column_ref: String,
        source: column_reference::Error,
    },
}

pub type Result<T> = std::result::Result<T, Error>;

#[derive(Debug, Clone, Copy, PartialEq, Default)]
pub enum Mode {
    #[default]
    Read,
    ReadWrite,
}

impl From<spicepod_dataset::Mode> for Mode {
    fn from(mode: spicepod_dataset::Mode) -> Self {
        match mode {
            spicepod_dataset::Mode::Read => Mode::Read,
            spicepod_dataset::Mode::ReadWrite => Mode::ReadWrite,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Default)]
pub enum TimeFormat {
    #[default]
    UnixSeconds,
    UnixMillis,
    ISO8601,
}

impl From<spicepod_dataset::TimeFormat> for TimeFormat {
    fn from(time_format: spicepod_dataset::TimeFormat) -> Self {
        match time_format {
            spicepod_dataset::TimeFormat::UnixSeconds => TimeFormat::UnixSeconds,
            spicepod_dataset::TimeFormat::UnixMillis => TimeFormat::UnixMillis,
            spicepod_dataset::TimeFormat::ISO8601 => TimeFormat::ISO8601,
        }
    }
}

impl std::fmt::Display for TimeFormat {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{self:?}")
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Dataset {
    pub from: String,
    pub name: TableReference,
    pub mode: Mode,
    pub params: HashMap<String, String>,
    pub has_metadata_table: bool,
    pub replication: Option<replication::Replication>,
    pub time_column: Option<String>,
    pub time_format: Option<TimeFormat>,
    pub acceleration: Option<acceleration::Acceleration>,
    pub embeddings: Vec<ColumnEmbeddingConfig>,
}

impl TryFrom<spicepod_dataset::Dataset> for Dataset {
    type Error = crate::Error;

    fn try_from(dataset: spicepod_dataset::Dataset) -> std::result::Result<Self, Self::Error> {
        let acceleration = dataset
            .acceleration
            .map(acceleration::Acceleration::try_from)
            .transpose()?;

        let table_reference = Dataset::parse_table_reference(&dataset.name)?;

        Ok(Dataset {
            from: dataset.from,
            name: table_reference,
            mode: Mode::from(dataset.mode),
            params: dataset
                .params
                .as_ref()
                .map(Params::as_string_map)
                .unwrap_or_default(),
            has_metadata_table: dataset
                .has_metadata_table
                .unwrap_or(Dataset::have_metadata_table_by_default()),
            replication: dataset.replication.map(replication::Replication::from),
            time_column: dataset.time_column,
            time_format: dataset.time_format.map(TimeFormat::from),
            embeddings: dataset.embeddings,
            acceleration,
        })
    }
}

impl Dataset {
    pub fn try_new(from: String, name: &str) -> std::result::Result<Self, crate::Error> {
        Ok(Dataset {
            from,
            name: Self::parse_table_reference(name)?,
            mode: Mode::default(),
            params: HashMap::default(),
            has_metadata_table: Self::have_metadata_table_by_default(),
            replication: None,
            time_column: None,
            time_format: None,
            acceleration: None,
            embeddings: Vec::default(),
        })
    }

    #[must_use]
    /// Returns whether the dataset should enable metadata by default.
    fn have_metadata_table_by_default() -> bool {
        false
    }

    pub(crate) fn parse_table_reference(
        name: &str,
    ) -> std::result::Result<TableReference, crate::Error> {
        match TableReference::parse_str(name) {
            table_ref @ (TableReference::Bare { .. } | TableReference::Partial { .. }) => {
                Ok(table_ref)
            }
            TableReference::Full { catalog, .. } => crate::DatasetNameIncludesCatalogSnafu {
                catalog,
                name: name.to_string(),
            }
            .fail(),
        }
    }

    /// Returns the dataset source - the first part of the `from` field before the first `:`.
    ///
    /// # Examples
    ///
    /// ```
    /// use runtime::component::dataset::Dataset;
    ///
    /// let dataset = Dataset::new("foo:bar".to_string(), "bar".to_string());
    ///
    /// assert_eq!(dataset.source(), "foo".to_string());
    /// ```
    ///
    /// ```
    /// use runtime::component::dataset::Dataset;
    ///
    /// let dataset = Dataset::new("foo".to_string(), "bar".to_string());
    ///
    /// assert_eq!(dataset.source(), "spiceai");
    /// ```
    #[must_use]
    pub fn source(&self) -> String {
        let parts: Vec<&str> = self.from.splitn(2, ':').collect();
        if parts.len() > 1 {
            parts[0].to_string()
        } else {
            if self.from == "localhost" || self.from.is_empty() {
                return "localhost".to_string();
            }
            "spiceai".to_string()
        }
    }

    /// Returns the dataset path - the remainder of the `from` field after the first `:` or the whole string if no `:`.
    ///
    /// # Examples
    ///
    /// ```
    /// use crate::component::dataset::Dataset;
    ///
    /// let dataset = Dataset::new("foo:bar".to_string(), "bar".to_string());
    ///
    /// assert_eq!(dataset.path(), "bar".to_string());
    /// ```
    ///
    /// ```
    /// use crate::component::dataset::Dataset;
    ///
    /// let dataset = Dataset::new("foo".to_string(), "bar".to_string());
    ///
    /// assert_eq!(dataset.path(), "foo".to_string());
    /// ```
    #[must_use]
    pub fn path(&self) -> String {
        match self.from.find(':') {
            Some(index) => self.from[index + 1..].to_string(),
            None => self.from.clone(),
        }
    }

    #[must_use]
    pub fn engine_secret(&self) -> Option<String> {
        if let Some(acceleration) = &self.acceleration {
            return acceleration.engine_secret.clone();
        }

        None
    }

    #[must_use]
    pub fn refresh_check_interval(&self) -> Option<Duration> {
        if let Some(acceleration) = &self.acceleration {
            if let Some(refresh_check_interval) = &acceleration.refresh_check_interval {
                if let Ok(duration) = fundu::parse_duration(refresh_check_interval) {
                    return Some(duration);
                }
                tracing::warn!(
                    "Unable to parse refresh interval for dataset {}: {}",
                    self.name,
                    refresh_check_interval
                );
            }
        }

        None
    }

    pub fn retention_check_interval(&self) -> Option<Duration> {
        if let Some(acceleration) = &self.acceleration {
            if let Some(retention_check_interval) = &acceleration.retention_check_interval {
                if let Ok(duration) = fundu::parse_duration(retention_check_interval) {
                    return Some(duration);
                }
                tracing::warn!(
                    "Unable to parse retention check interval for dataset {}: {}",
                    self.name,
                    retention_check_interval
                );
            }
        }

        None
    }

    pub fn retention_period(&self) -> Option<Duration> {
        if let Some(acceleration) = &self.acceleration {
            if let Some(retention_period) = &acceleration.retention_period {
                if let Ok(duration) = fundu::parse_duration(retention_period) {
                    return Some(duration);
                }
                tracing::warn!(
                    "Unable to parse retention period for dataset {}: {}",
                    self.name,
                    retention_period
                );
            }
        }

        None
    }

    #[must_use]
    pub fn refresh_sql(&self) -> Option<String> {
        if let Some(acceleration) = &self.acceleration {
            return acceleration.refresh_sql.clone();
        }

        None
    }

    #[must_use]
    pub fn refresh_data_window(&self) -> Option<Duration> {
        if let Some(acceleration) = &self.acceleration {
            if let Some(refresh_data_window) = &acceleration.refresh_data_window {
                if let Ok(duration) = fundu::parse_duration(refresh_data_window) {
                    return Some(duration);
                }
                tracing::warn!(
                    "Unable to parse refresh period for dataset {}: {}",
                    self.name,
                    refresh_data_window
                );
            }
        }

        None
    }

    #[must_use]
    pub fn mode(&self) -> Mode {
        self.mode
    }

    #[must_use]
    pub fn is_accelerated(&self) -> bool {
        if let Some(acceleration) = &self.acceleration {
            return acceleration.enabled;
        }

        false
    }

    #[must_use]
    pub fn is_file_accelerated(&self) -> bool {
        if let Some(acceleration) = &self.acceleration {
            return acceleration.enabled && acceleration.mode == acceleration::Mode::File;
        }

        false
    }
}

pub mod acceleration;

pub mod replication {
    use spicepod::component::dataset::replication as spicepod_replication;

    #[derive(Debug, Clone, PartialEq, Default)]
    pub struct Replication {
        pub enabled: bool,
    }

    impl From<spicepod_replication::Replication> for Replication {
        fn from(replication: spicepod_replication::Replication) -> Self {
            Replication {
                enabled: replication.enabled,
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use std::collections::HashMap;

    use data_components::util::column_reference::ColumnReference;

    use super::acceleration::{Acceleration, IndexType};

    #[test]
    fn test_indexes_roundtrip() {
        let indexes_map = HashMap::from([
            ("foo".to_string(), IndexType::Enabled),
            ("bar".to_string(), IndexType::Unique),
        ]);

        let indexes_str = Acceleration::hashmap_to_option_string(&indexes_map);
        assert!(indexes_str == "foo:enabled;bar:unique" || indexes_str == "bar:unique;foo:enabled");
        let roundtrip_indexes_map: HashMap<String, IndexType> =
            data_components::util::hashmap_from_option_string(&indexes_str);

        let roundtrip_indexes_map = roundtrip_indexes_map
            .into_iter()
            .map(|(k, v)| (k, v.to_string()))
            .collect::<HashMap<String, String>>();

        let indexes_map = indexes_map
            .into_iter()
            .map(|(k, v)| (k, v.to_string()))
            .collect::<HashMap<String, String>>();

        assert_eq!(indexes_map, roundtrip_indexes_map);
    }

    #[test]
    fn test_compound_indexes_roundtrip() {
        let indexes_map = HashMap::from([
            ("(foo, bar)".to_string(), IndexType::Enabled),
            ("bar".to_string(), IndexType::Unique),
        ]);

        let indexes_str = Acceleration::hashmap_to_option_string(&indexes_map);
        assert!(
            indexes_str == "(foo, bar):enabled;bar:unique"
                || indexes_str == "bar:unique;(foo, bar):enabled"
        );
        let roundtrip_indexes_map: HashMap<String, IndexType> =
            data_components::util::hashmap_from_option_string(&indexes_str);

        let roundtrip_indexes_map = roundtrip_indexes_map
            .into_iter()
            .map(|(k, v)| (k, v.to_string()))
            .collect::<HashMap<String, String>>();

        let indexes_map = indexes_map
            .into_iter()
            .map(|(k, v)| (k, v.to_string()))
            .collect::<HashMap<String, String>>();

        assert_eq!(indexes_map, roundtrip_indexes_map);
    }

    #[test]
    fn test_get_index_columns() {
        let column_ref = ColumnReference::try_from("foo").expect("valid columns");
        assert_eq!(column_ref.iter().collect::<Vec<_>>(), vec!["foo"]);

        let column_ref = ColumnReference::try_from("(foo, bar)").expect("valid columns");
        assert_eq!(column_ref.iter().collect::<Vec<_>>(), vec!["bar", "foo"]);

        let column_ref = ColumnReference::try_from("(foo,bar)").expect("valid columns");
        assert_eq!(column_ref.iter().collect::<Vec<_>>(), vec!["bar", "foo"]);

        let err = ColumnReference::try_from("(foo,bar").expect_err("invalid columns");
        assert_eq!(
            err.to_string(),
            "The column reference \"(foo,bar\" is missing a closing parenthensis."
        );
    }
}
