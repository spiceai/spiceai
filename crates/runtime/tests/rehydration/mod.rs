/*
Copyright 2024 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
//! Runs federation integration tests for `MySQL`.
//!
//! Expects a Docker daemon to be running.
use crate::{
    docker::RunningContainer,
    get_test_datafusion,
    mysql::common::{get_mysql_conn, make_mysql_dataset, start_mysql_docker_container},
    runtime_ready_check,
};
use std::sync::Arc;

use crate::init_tracing;

use anyhow::Context;
use app::AppBuilder;
use arrow::array::RecordBatch;
use datafusion_table_providers::sql::arrow_sql_gen::statement::{
    CreateTableBuilder, InsertBuilder,
};
use futures::TryStreamExt;
use mysql_async::{prelude::Queryable, Params, Row};
use runtime::{datafusion::query::Protocol, spice_data_base_path, status, Runtime};
use spicepod::component::dataset::{
    acceleration::{Acceleration, Mode},
    Dataset,
};

use spicepod::component::params::Params as SpicepodParams;

use tracing::instrument;

#[tokio::test]
async fn spill_to_disk_and_rehydration() -> Result<(), anyhow::Error> {
    let running_container = prepare_test_environment()
        .await
        .map_err(|e| anyhow::anyhow!(e))?;
    let running_container = Arc::new(running_container);

    let config = vec![
        // #[cfg(feature = "duckdb")]
        // ("duckdb", None),
        // #[cfg(feature = "duckdb")]
        // ("duckdb", Some("./.spice/my_duckdb.db")),
        #[cfg(feature = "sqlite")]
        ("sqlite", None),
        #[cfg(feature = "sqlite")]
        ("sqlite", Some("./.spice/my_sqlite.db")),
    ];

    for (idx, (engine, db_file_path)) in config.into_iter().enumerate() {
        tracing::info!("Testing spill-to-disk and rehydration with engine: {engine}, db_file_path: {db_file_path:?}");

        if idx > 0 {
            // Ensure the container is running as the tests manipulate the container
            running_container
                .start()
                .await
                .map_err(|e| anyhow::anyhow!(e))?;
            tokio::time::sleep(std::time::Duration::from_millis(5000)).await;
        }
        execute_spill_to_disk_and_rehydration(Arc::clone(&running_container), engine, db_file_path)
            .await?;
    }

    running_container
        .remove()
        .await
        .map_err(|e| anyhow::anyhow!(e))?;

    tracing::info!("Spill-to-disk and rehydration tests passed!");

    Ok(())
}

/// Validates spill-to-disk and rehydration functionality by simulating runtime restarts
/// and checking data consistency.
///
/// 1. Retrieve the number of rows using a native `MySQL` connection to use as a baseline.
/// 2. Start Spice, retrieve row count and loaded items after acceleration is completed, and compare with the baseline.
/// 3. Restart the runtime and ensure the loaded items remain consistent immediately after the runtime is loaded.
/// 4. Simulate federated dataset access issue after the runtime is restarted, ensure query result remain consistent.
///
async fn execute_spill_to_disk_and_rehydration(
    federated_dataset_container: Arc<RunningContainer<'static>>,
    engine: &str,
    db_file_path: Option<&str>,
) -> Result<(), anyhow::Error> {
    // retrieve number of rows using native mysql connection
    // this also ensures that federated dataset is available
    let pool = get_mysql_conn(MYSQL_PORT)?;
    let res: Vec<Row> = pool
        .get_conn()
        .await?
        .exec("SELECT COUNT(*) FROM lineitem", Params::Empty)
        .await?;
    let num_rows: u64 = res[0].get(0).context("Unable to retrieve number of rows")?;

    let accelerated_db_file_path = resolve_local_db_file_path(engine, db_file_path)?;
    tracing::debug!(
        "Expected accelerated database location: {}",
        &accelerated_db_file_path
    );

    // clean up: delete local database file if exists before running the test
    for file_path in [
        accelerated_db_file_path.clone(),
        format!("{accelerated_db_file_path}-wal"),
        format!("{accelerated_db_file_path}-shm"),
    ] {
        if std::fs::metadata(&file_path).is_ok() {
            std::fs::remove_file(&file_path).context("should remove local database")?;
        }
    }

    let rt = init_spice_app(engine, db_file_path).await?;
    runtime_ready_check(&rt).await;

    if std::fs::metadata(&accelerated_db_file_path).is_err() {
        return Err(anyhow::anyhow!(
            "Accelerated database file not found at path: {accelerated_db_file_path}"
        ));
    }

    let test_query =
        "SELECT l_orderkey, l_linenumber  FROM lineitem ORDER BY l_orderkey, l_linenumber";

    let original_items = run_query(test_query, &rt).await?;
    let num_rows_loaded: usize = original_items
        .iter()
        .map(arrow::array::RecordBatch::num_rows)
        .sum();

    // ensure data has been loaded correctly
    assert_eq!(num_rows_loaded as u64, num_rows);

    drop(rt);

    // Restart the runtime and ensure the loaded items remain consistent
    let rt = init_spice_app(engine, db_file_path).await?;
    // Do request immediately after restart w/o waiting for ready status (dataset is refreshed)
    let restart1_items = run_query(test_query, &rt).await?;
    assert_eq!(original_items, restart1_items);

    drop(rt);

    // Simulate federated dataset access issue after the runtime is restarted, ensure query result remain consistent
    let rt = init_spice_app(engine, db_file_path).await?;
    federated_dataset_container.stop().await?;
    let restart2_items = run_query(test_query, &rt).await?;
    assert_eq!(original_items, restart2_items);

    Ok(())
}

fn resolve_local_db_file_path(
    engine: &str,
    db_file_path: Option<&str>,
) -> Result<String, anyhow::Error> {
    if let Some(db_file_path) = db_file_path {
        let working_dir = std::env::current_dir().unwrap_or(".".into());
        return Ok(format!(
            "{}/{db_file_path}",
            working_dir.to_str().context("Unable to get current dir")?
        ));
    }

    match engine {
        "duckdb" => Ok(format!("{}/lineitem.db", spice_data_base_path())),
        _ => Ok(format!("{}/lineitem_{engine}.db", spice_data_base_path())),
    }
}

async fn run_query(query: &str, rt: &Runtime) -> Result<Vec<RecordBatch>, anyhow::Error> {
    let query_result = rt
        .datafusion()
        .query_builder(query, Protocol::Internal)
        .build()
        .run()
        .await
        .map_err(|e| anyhow::anyhow!("Failed to run query: {:?}", e))?;

    let collected_data = query_result
        .data
        .try_collect::<Vec<_>>()
        .await
        .map_err(|e| anyhow::anyhow!("Failed to collect query results: {:?}", e))?;

    Ok(collected_data)
}

async fn init_spice_app(
    acceleration_engine: &str,
    db_file_path: Option<&str>,
) -> Result<Runtime, anyhow::Error> {
    let ds = create_test_dataset(acceleration_engine, db_file_path);

    let app = AppBuilder::new("spiceapp").with_dataset(ds).build();

    let status = status::RuntimeStatus::new();
    let df = get_test_datafusion(Arc::clone(&status));

    let rt = Runtime::builder()
        .with_app(app)
        .with_datafusion(df)
        .with_runtime_status(status)
        .build()
        .await;

    tokio::select! {
        () = tokio::time::sleep(std::time::Duration::from_secs(10)) => {
            return Err(anyhow::anyhow!("Timed out waiting for datasets to load"));
        }
        () = rt.load_components() => {}
    }

    Ok(rt)
}

fn create_test_dataset(acceleration_engine: &str, db_file_path: Option<&str>) -> Dataset {
    let mut ds = make_mysql_dataset("lineitem", "lineitem", MYSQL_PORT, false);

    let mut acceleration = Acceleration {
        enabled: true,
        engine: Some(acceleration_engine.to_string()),
        mode: Mode::File,
        ..Default::default()
    };

    if let Some(db_file_path) = db_file_path {
        let params = SpicepodParams::from_string_map(
            vec![(
                format!("{acceleration_engine}_file",),
                db_file_path.to_string(),
            )]
            .into_iter()
            .collect(),
        );
        acceleration.params = Some(params);
    }

    ds.acceleration = Some(acceleration);

    ds
}

const MYSQL_DOCKER_CONTAINER: &str = "runtime-integration-test-rehydration-retry-mysql";
const MYSQL_PORT: u16 = 13337;

#[instrument]
async fn init_mysql_db() -> Result<(), anyhow::Error> {
    let pool = get_mysql_conn(MYSQL_PORT)?;
    let mut conn = pool.get_conn().await?;

    tracing::debug!("DROP TABLE IF EXISTS lineitem");
    let _: Vec<Row> = conn
        .exec("DROP TABLE IF EXISTS lineitem", Params::Empty)
        .await?;

    tracing::debug!("Downloading TPCH lineitem...");
    let tpch_lineitem = crate::get_tpch_lineitem().await?;

    let tpch_lineitem_schema = Arc::clone(&tpch_lineitem[0].schema());

    let create_table_stmt = CreateTableBuilder::new(tpch_lineitem_schema, "lineitem").build_mysql();
    tracing::debug!("CREATE TABLE lineitem...");
    let _: Vec<Row> = conn.exec(create_table_stmt, Params::Empty).await?;

    tracing::debug!("INSERT INTO lineitem...");
    let insert_stmt = InsertBuilder::new("lineitem", tpch_lineitem).build_mysql(None)?;
    let _: Vec<Row> = conn.exec(insert_stmt, Params::Empty).await?;
    tracing::debug!("MySQL initialized!");

    Ok(())
}

#[instrument]
async fn prepare_test_environment() -> Result<RunningContainer<'static>, String> {
    let _tracing = init_tracing(Some("integration=debug,info"));
    let running_container = start_mysql_docker_container(MYSQL_DOCKER_CONTAINER, MYSQL_PORT)
        .await
        .map_err(|e| {
            tracing::error!("Failed to start MySQL Docker container: {e}");
            e.to_string()
        })?;
    tracing::debug!("Container started");
    init_mysql_db().await.map_err(|e| {
        tracing::error!("Failed to initialize MySQL database: {e}");
        e.to_string()
    })?;

    Ok(running_container)
}
