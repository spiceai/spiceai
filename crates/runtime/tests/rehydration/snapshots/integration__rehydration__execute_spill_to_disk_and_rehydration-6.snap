---
source: crates/runtime/tests/rehydration/mod.rs
expression: restart2_items_pretty
---
+------------+--------------+
| l_orderkey | l_linenumber |
+------------+--------------+
| 1          | 2            |
| 1          | 3            |
| 1          | 4            |
| 1          | 5            |
| 1          | 6            |
| 2          | 1            |
| 3          | 1            |
| 3          | 2            |
| 3          | 3            |
| 3          | 4            |
+------------+--------------+
